modemController.go

```diff
@@ -0,0 +1,149 @@
+	"time"
+
+	"periph.io/x/periph/conn/gpio"
+	"periph.io/x/periph/conn/gpio/gpioreg"
+)
+
+type ModemController struct {
+	StartTime         time.Time
+	Modem             *Modem
+	ModemsConfig      []ModemConfig
+	TestHosts         []string
+	TestInterval      int
+	PowerPin          string
+	InitialOnTime     int
+	FindModemTime     int // Time in seconds after USB powered on for the modem to be found
+	ConnectionTimeout int // Time in seconds for modem to make a connection to the network
+	PingWaitTime      int
+	PingRetries       int
+	RequestOnTime     int // Time the modem will stay on in seconds after a request was made
+
+	lastOnRequestTime time.Time
+}
+
+func (mc *ModemController) NewOnRequest() {
+	log.Println("stay on request")
+	mc.lastOnRequestTime = time.Now()
+}
+
+func (mc *ModemController) FindModem() bool {
+	startTime := time.Now()
+	for {
+		for _, modemConfig := range mc.ModemsConfig {
+			cmd := exec.Command("lsusb", "-d", modemConfig.VendorProduct)
+			if err := cmd.Run(); err == nil {
+				mc.Modem = NewModem(modemConfig)
+				return true
+			}
+		}
+		if timeoutCheck(startTime, mc.FindModemTime) {
+			return false
+		}
+		time.Sleep(time.Second)
+	}
+}
+
+func (mc *ModemController) SetModemPower(on bool) error {
+	pin := gpioreg.ByName(mc.PowerPin)
+	if on {
+		if err := pin.Out(gpio.High); err != nil {
+			return fmt.Errorf("failed to set modem power pin high: %v", err)
+		}
+	} else {
+		if err := pin.Out(gpio.Low); err != nil {
+			return fmt.Errorf("failed to set modem power pin low: %v", err)
+		}
+	}
+	return nil
+}
+
+func (mc *ModemController) CycleModemPower() error {
+	if err := mc.SetModemPower(false); err != nil {
+		return err
+	}
+	time.Sleep(time.Second * 5)
+	return mc.SetModemPower(true)
+}
+
+// WaitForConnection will return false if no connection is made before either
+// it timeouts or the modem should no longer be powered.
+func (mc *ModemController) WaitForConnection() (bool, error) {
+	startTime := time.Now()
+	for {
+		def, err := mc.Modem.IsDefaultRoute()
+		if err != nil {
+			return false, err
+		}
+		if def {
+			return true, nil
+		}
+		if timeoutCheck(startTime, mc.ConnectionTimeout) {
+			return false, nil
+		}
+		time.Sleep(time.Second)
+	}
+	return false, nil
+}
+
+// ShouldBeOff will look at the following factors to determine if the modem shoudl be off.
+// - InitialOnTime: Modem shoudl be on for a set amount of time at the start.
+// - LastOnRequest: //TODO
+// - OnWindow: //TODO
+func (mc *ModemController) ShouldBeOff() bool {
+	if !timeoutCheck(mc.StartTime, mc.InitialOnTime) {
+		return false
+	}
+
+	if !timeoutCheck(mc.lastOnRequestTime, mc.RequestOnTime) {
+		return false
+	}
+
+	return true
+}
+
+// WaitForNextPingTest will return false if when waiting ShouldBeOff returns
+// true, otherwise will return true after waiting.
+func (mc *ModemController) WaitForNextPingTest() bool {
+	startTime := time.Now()
+	for {
+		if timeoutCheck(startTime, mc.TestInterval) {
+			return true
+		}
```

Comment:
A nicer way to structure this would be:

```go
timeout := time.After(mc.TestInterval)
for {
    select {
        case <-timeout:
            return true
        case <-time.After(time.Second):
            if mc.ShouldBeOff() {
                return false
            }
        }
     }
}
```

this lets time.After handle the time calculations, avoiding needing to track the start time.